### PR TITLE
use rufus-scheduler instead of /api/rest/system/job/:id/run endpoint

### LINKED ISSUE #986: Migrate cron tasks to rufus-scheduler
https://github.com/jmettraux/rufus-scheduler

scheduler should be started as dedicated process(not  inside web server), see https://github.com/jmettraux/rufus-scheduler/issues/307

### PR DESCRIPTION
**Pull request by senid231**
fixes #986

example of process log

```
$ RAILS_LOG_TO_STDOUT=true bundle exec bin/cron_scheduler
App Timezone Europe/Kiev
   (0.2ms)  SELECT current_setting('TIMEZONE');
  ↳ /home/senid/.rvm/gems/ruby-2.6.5@yeti-web/bundler/gems/pg_sql_caller-c87b59db82ee/lib/pg_sql_caller/base.rb:44
Yeti DB timezone UTC
   (0.2ms)  SELECT current_setting('TIMEZONE');
  ↳ /home/senid/.rvm/gems/ruby-2.6.5@yeti-web/bundler/gems/pg_sql_caller-c87b59db82ee/lib/pg_sql_caller/base.rb:44
CDR DB timezone UTC
ActiveRecord timezone utc
Raven 2.12.3 configured not to capture errors: No host specified, no public_key specified, no project_id specified
CronScheduler process started.
[CronScheduler] Started cron job DialpeerRatesApply.
[CronScheduler] Started cron job CallsMonitoring.
[CronScheduler]    (1.4ms)  BEGIN
[CronScheduler]   ↳ app/jobs/base_job.rb:35
[CronScheduler]    (1.5ms)  BEGIN
[CronScheduler]   ↳ app/jobs/base_job.rb:35
[CronScheduler] Started cron job EventProcessor.
[CronScheduler] Started cron job AccountBalanceNotify.
[CronScheduler]    (0.9ms)  BEGIN
[CronScheduler]   ↳ app/jobs/base_job.rb:35
[CronScheduler]    (0.6ms)  BEGIN
[CronScheduler]   ↳ app/jobs/base_job.rb:35
[CronScheduler]   BaseJob Load (0.8ms)  SELECT  "sys"."jobs".* FROM "sys"."jobs" WHERE "sys"."jobs"."running" = $1 AND ("sys"."jobs"."id" = 0 OR "sys"."jobs"."type" = 'DialpeerRatesApply') ORDER BY "sys"."jobs"."id" ASC LIMIT $2 FOR UPDATE NOWAIT  [["running", false], ["LIMIT", 1]]
[CronScheduler]   ↳ app/jobs/base_job.rb:39
[CronScheduler] Starting scheduler DialpeerRatesApply
[CronScheduler]   BaseJob Load (7.6ms)  SELECT  "sys"."jobs".* FROM "sys"."jobs" WHERE "sys"."jobs"."running" = $1 AND ("sys"."jobs"."id" = 0 OR "sys"."jobs"."type" = 'CallsMonitoring') ORDER BY "sys"."jobs"."id" ASC LIMIT $2 FOR UPDATE NOWAIT  [["running", false], ["LIMIT", 1]]
[CronScheduler]   ↳ app/jobs/base_job.rb:39
[CronScheduler]   Jobs::DialpeerRatesApply Update (0.8ms)  UPDATE "sys"."jobs" SET "running" = $1 WHERE "sys"."jobs"."id" = $2  [["running", true], ["id", 11]]
[CronScheduler]   ↳ app/jobs/base_job.rb:17
[CronScheduler]   BaseJob Load (8.8ms)  SELECT  "sys"."jobs".* FROM "sys"."jobs" WHERE "sys"."jobs"."running" = $1 AND ("sys"."jobs"."id" = 0 OR "sys"."jobs"."type" = 'AccountBalanceNotify') ORDER BY "sys"."jobs"."id" ASC LIMIT $2 FOR UPDATE NOWAIT  [["running", false], ["LIMIT", 1]]
[CronScheduler]   ↳ app/jobs/base_job.rb:39
[CronScheduler]    (0.4ms)  COMMIT
[CronScheduler]   BaseJob Load (1.6ms)  SELECT  "sys"."jobs".* FROM "sys"."jobs" WHERE "sys"."jobs"."running" = $1 AND ("sys"."jobs"."id" = 0 OR "sys"."jobs"."type" = 'EventProcessor') ORDER BY "sys"."jobs"."id" ASC LIMIT $2 FOR UPDATE NOWAIT  [["running", false], ["LIMIT", 1]]
[CronScheduler]   ↳ app/jobs/base_job.rb:39
[CronScheduler]   ↳ app/jobs/base_job.rb:35
[CronScheduler]    (0.1ms)  BEGIN
[CronScheduler]   ↳ app/jobs/base_job.rb:22
[CronScheduler] Starting scheduler CallsMonitoring
[CronScheduler]   Jobs::CallsMonitoring Update (1.1ms)  UPDATE "sys"."jobs" SET "running" = $1 WHERE "sys"."jobs"."id" = $2  [["running", true], ["id", 5]]
[CronScheduler]   ↳ app/jobs/base_job.rb:17
[CronScheduler] Starting scheduler EventProcessor
[CronScheduler]    (0.7ms)  COMMIT
[CronScheduler]   ↳ app/jobs/base_job.rb:35
[CronScheduler] Starting scheduler AccountBalanceNotify
[CronScheduler]   Jobs::EventProcessor Update (0.5ms)  UPDATE "sys"."jobs" SET "running" = $1 WHERE "sys"."jobs"."id" = $2  [["running", true], ["id", 2]]
[CronScheduler]   ↳ app/jobs/base_job.rb:17
[CronScheduler]   Jobs::AccountBalanceNotify Update (0.3ms)  UPDATE "sys"."jobs" SET "running" = $1 WHERE "sys"."jobs"."id" = $2  [["running", true], ["id", 12]]
[CronScheduler]   ↳ app/jobs/base_job.rb:17
[CronScheduler]   DialpeerNextRate Load (1.0ms)  SELECT  "dialpeer_next_rates".* FROM "dialpeer_next_rates" WHERE "dialpeer_next_rates"."applied" = $1 AND (apply_time < '2021-05-31 16:14:00.185638') ORDER BY "dialpeer_next_rates"."id" ASC LIMIT $2  [["applied", false], ["LIMIT", 1000]]
[CronScheduler]   ↳ app/jobs/jobs/dialpeer_rates_apply.rb:7
[CronScheduler]    (0.4ms)  BEGIN
[CronScheduler]   ↳ app/jobs/base_job.rb:22
[CronScheduler]    (1.5ms)  COMMIT
[CronScheduler]   ↳ app/jobs/base_job.rb:35
[CronScheduler]    (1.1ms)  COMMIT
[CronScheduler]   ↳ app/jobs/base_job.rb:35
[CronScheduler]    (0.2ms)  BEGIN
[CronScheduler]   ↳ app/jobs/base_job.rb:22
[CronScheduler]   Node Load (1.5ms)  SELECT "nodes".* FROM "nodes"
[CronScheduler]    (0.9ms)  BEGIN
[CronScheduler]   ↳ app/jobs/base_job.rb:22
[CronScheduler]   ↳ app/models/yeti/cdrs_filter.rb:40
[CronScheduler]   Log::BalanceNotification Load (0.7ms)  SELECT "balance_notifications".* FROM "balance_notifications" WHERE (not is_processed) ORDER BY created_at
[CronScheduler]   ↳ app/jobs/jobs/account_balance_notify.rb:6
[CronScheduler]    (0.5ms)  SELECT COUNT(*) FROM "events"
[CronScheduler]   ↳ app/jobs/jobs/event_processor.rb:7
[CronScheduler] Scoped order is ignored, it's forced to be batch order.
[CronScheduler]    (1.8ms)  SELECT COUNT(*) FROM "nodes"
[CronScheduler]   ↳ app/models/yeti/cdrs_filter.rb:40
[CronScheduler]  total calls count: 0 
[CronScheduler]   Routing::DestinationNextRate Load (0.7ms)  SELECT  "class4"."destination_next_rates".* FROM "class4"."destination_next_rates" WHERE "class4"."destination_next_rates"."applied" = $1 AND (apply_time < '2021-05-31 16:14:00.195715') ORDER BY "class4"."destination_next_rates"."id" ASC LIMIT $2  [["applied", false], ["LIMIT", 1000]]
[CronScheduler]   ↳ app/jobs/jobs/dialpeer_rates_apply.rb:21
[CronScheduler]    (1.0ms)  SELECT COUNT(*) FROM "balance_notifications" WHERE (not is_processed)
[CronScheduler]   ↳ app/jobs/jobs/account_balance_notify.rb:14
[CronScheduler]    (0.2ms)  COMMIT
[CronScheduler]   ↳ app/jobs/base_job.rb:22
[CronScheduler]    (0.4ms)  COMMIT
[CronScheduler]   ↳ app/jobs/base_job.rb:22
[CronScheduler]   Jobs::DialpeerRatesApply Update (0.6ms)  UPDATE "sys"."jobs" SET "running" = $1, "updated_at" = $2 WHERE "sys"."jobs"."id" = $3  [["running", false], ["updated_at", "2021-05-31 16:14:00.201447"], ["id", 11]]
[CronScheduler]   ↳ app/jobs/base_job.rb:57
[CronScheduler] Cron job DialpeerRatesApply was performed successfully.
[CronScheduler]   Jobs::AccountBalanceNotify Update (0.3ms)  UPDATE "sys"."jobs" SET "running" = $1, "updated_at" = $2 WHERE "sys"."jobs"."id" = $3  [["running", false], ["updated_at", "2021-05-31 16:14:00.201899"], ["id", 12]]
[CronScheduler]   ↳ app/jobs/base_job.rb:57
[CronScheduler] Cron job AccountBalanceNotify was performed successfully.
[CronScheduler]   Event Load (0.2ms)  SELECT  "events".* FROM "events" ORDER BY "events"."id" ASC LIMIT $1  [["LIMIT", 1000]]
[CronScheduler]   ↳ app/jobs/jobs/event_processor.rb:8
[CronScheduler] Job EventProcessor processed 0/0 events.
[CronScheduler]    (0.1ms)  COMMIT
[CronScheduler]   ↳ app/jobs/base_job.rb:22
[CronScheduler]   GuiConfig Load (0.4ms)  SELECT  "guiconfig".* FROM "guiconfig" ORDER BY "guiconfig"."id" ASC LIMIT $1  [["LIMIT", 1]]
[CronScheduler]   ↳ app/models/gui_config.rb:66
[CronScheduler]   Jobs::EventProcessor Update (0.3ms)  UPDATE "sys"."jobs" SET "running" = $1, "updated_at" = $2 WHERE "sys"."jobs"."id" = $3  [["running", false], ["updated_at", "2021-05-31 16:14:00.205565"], ["id", 2]]
[CronScheduler]   ↳ app/jobs/base_job.rb:57
[CronScheduler] Cron job EventProcessor was performed successfully.
[CronScheduler]    (0.1ms)  BEGIN
[CronScheduler]   ↳ app/jobs/jobs/calls_monitoring.rb:384
[CronScheduler]    (0.1ms)  SELECT "nodes"."id" FROM "nodes"
[CronScheduler]   ↳ app/models/stats/base.rb:7
[CronScheduler]    (0.5ms)  SELECT "accounts"."id" FROM "accounts" LEFT OUTER JOIN "contractors" ON "contractors"."id" = "accounts"."contractor_id"
[CronScheduler]   ↳ app/models/stats/active_call_account.rb:35
[CronScheduler]   Stats::ActiveCallAccount Create (0.2ms)  INSERT INTO "stats"."active_call_accounts" ("account_id", "originated_count", "terminated_count", "created_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["account_id", 21], ["originated_count", 0], ["terminated_count", 0], ["created_at", "2021-05-31 16:14:00.191692"]]
[CronScheduler]   ↳ app/models/stats/active_call_account.rb:47
[CronScheduler]   Stats::ActiveCallAccount Create (0.1ms)  INSERT INTO "stats"."active_call_accounts" ("account_id", "originated_count", "terminated_count", "created_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["account_id", 20], ["originated_count", 0], ["terminated_count", 0], ["created_at", "2021-05-31 16:14:00.191692"]]
[CronScheduler]   ↳ app/models/stats/active_call_account.rb:47
[CronScheduler]    (0.5ms)  SELECT "gateways"."id" FROM "gateways"
[CronScheduler]   ↳ app/models/stats/base.rb:7
[CronScheduler]   Stats::ActiveCallOrigGateway Create (0.2ms)  INSERT INTO "stats"."active_call_orig_gateways" ("gateway_id", "count", "created_at") VALUES ($1, $2, $3) RETURNING "id"  [["gateway_id", 16], ["count", 0], ["created_at", "2021-05-31 16:14:00.191692"]]
[CronScheduler]   ↳ app/models/stats/base.rb:18
[CronScheduler]   Stats::ActiveCallOrigGateway Create (0.1ms)  INSERT INTO "stats"."active_call_orig_gateways" ("gateway_id", "count", "created_at") VALUES ($1, $2, $3) RETURNING "id"  [["gateway_id", 17], ["count", 0], ["created_at", "2021-05-31 16:14:00.191692"]]
[CronScheduler]   ↳ app/models/stats/base.rb:18
[CronScheduler]    (0.1ms)  SELECT "gateways"."id" FROM "gateways"
[CronScheduler]   ↳ app/models/stats/base.rb:7
[CronScheduler]   Stats::ActiveCallTermGateway Create (0.2ms)  INSERT INTO "stats"."active_call_term_gateways" ("gateway_id", "count", "created_at") VALUES ($1, $2, $3) RETURNING "id"  [["gateway_id", 16], ["count", 0], ["created_at", "2021-05-31 16:14:00.191692"]]
[CronScheduler]   ↳ app/models/stats/base.rb:18
[CronScheduler]   Stats::ActiveCallTermGateway Create (0.1ms)  INSERT INTO "stats"."active_call_term_gateways" ("gateway_id", "count", "created_at") VALUES ($1, $2, $3) RETURNING "id"  [["gateway_id", 17], ["count", 0], ["created_at", "2021-05-31 16:14:00.191692"]]
[CronScheduler]   ↳ app/models/stats/base.rb:18
[CronScheduler]    (0.1ms)  COMMIT
[CronScheduler]   ↳ app/jobs/jobs/calls_monitoring.rb:384
[CronScheduler]   Node Load (0.1ms)  SELECT "nodes".* FROM "nodes"
[CronScheduler]   ↳ app/jobs/jobs/calls_monitoring.rb:395
[CronScheduler]    (0.1ms)  COMMIT
[CronScheduler]   ↳ app/jobs/base_job.rb:22
[CronScheduler]   Jobs::CallsMonitoring Update (0.1ms)  UPDATE "sys"."jobs" SET "running" = $1, "updated_at" = $2 WHERE "sys"."jobs"."id" = $3  [["running", false], ["updated_at", "2021-05-31 16:14:00.257810"], ["id", 5]]
[CronScheduler]   ↳ app/jobs/base_job.rb:57
[CronScheduler] Cron job CallsMonitoring was performed successfully.
```

### CODE REVIEW

File: lib/cron_scheduler.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    # each time thread should acquire and release database connection.
+    Yeti::ActiveRecord.clear_active_connections!
+    Yeti::ActiveRecord.flush_idle_connections!
+    Cdr::Base.clear_active_connections!
+    Cdr::Base.flush_idle_connections!
+
+    self._current_scheduler = Rufus::Scheduler.new
+
+    jobs.each do |job|
+      _current_scheduler.cron(job.cron_time) { new(job).run! }
```
Comment (Fivell) [inline lib/cron_scheduler.rb:44]: :overlap => false option should be added

File: lib/cron_scheduler.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    BaseJob.launch!(job_class_name)
+    logger.info { "Cron job #{job_class_name} was performed successfully." }
+  rescue StandardError => e
+    logger.error { "Cron job #{job_class_name} was performed with an exception." }
+    logger.error { "#{e.class}>: #{e.message}\n#{e.backtrace&.join("\n")}" }
+  end
+
+  def wrap_connections
+    Yeti::ActiveRecord.connection_pool.with_connection do
+      Cdr::Base.connection_pool.with_connection { yield }
```
Comment (Fivell) [inline lib/cron_scheduler.rb:78]: I think database connection pool should be reconfigured since up to more 13 connections can be taken by rufus threads

Comment (senid231) [reply lib/cron_scheduler.rb:78]: it will be configured in database.yml on production server

File: lib/cron_scheduler.rb (hunk 1 above)
Comment (senid231) [reply lib/cron_scheduler.rb:44]: will be added in separate PR, during replace `BaseJob` AR model with PORO ruby object

**Pull request merged**